site/src/sphinx/client-service-discovery.rst

```diff
@@ -0,0 +1,237 @@
+
+.. _creating-endpoint-group:
+
+
+Creating an ``EndpointGroup``
+-----------------------------
+There are various :api:`EndpointGroup` implementations provided out of the box, but let's start simple with
+:api:`StaticEndpointGroup` which always yields a pre-defined set of :apiplural:`Endpoint` specified
+at construction time:
+
+.. code-block:: java
+
+    // Create a group of well-known search engine endpoints.
+    EndpointGroup searchEngineGroup = new StaticEndpointGroup(
+            Endpoint.of("www.google.com", 443),
+            Endpoint.of("www.bing.com", 443),
+            Endpoint.of("duckduckgo.com", 443);
+
+    List<Endpoint> endpoints = searchEngineGroup.endpoints();
+    assert endpoints.contains(Endpoint.of("www.google.com", 443));
+    assert endpoints.contains(Endpoint.of("www.bing.com", 443));
+    assert endpoints.contains(Endpoint.of("duckduckgo.com", 443));
+
+
+Registering an ``EndointGroup``
+-------------------------------
+An :api:`EndpointGroup` becomes visible by a client such as :api:`HttpClient` only after it's registered in
+:api:`EndpointGroupRegistry`. You need to specify 2 more elements to register an :api:`EndpointGroup`:
+
+- The name of the :api:`EndpointGroup`
+- An :api:`EndpointSelectionStrategy` that determines which :api:`Endpoint` is used for each request
+
+  - Use ``EndpointSelectionStrategy.WEIGHTED_ROUND_ROBIN`` for weighted round robin.
+  - Use ``EndpointSelectionStrategy.ROUND_ROBIN`` for unweighted rounb robin.
+  - Use :api:`StickyEndpointSelectionStrategy` if you want to pin the requests based on a criteria
+    such as a request parameter value.
+  - You can implement your own :api:`EndpointSelectionStrategy`.
+
+The following example registers the ``searchEngineGroup`` we created at :ref:`creating-endpoint-group`:
+
+.. code-block:: java
+
+    EndpointGroupRegistry.register("search_engines", searchEngineGroup,
+                                   EndpointSelectionStrategy.WEIGHTED_ROUND_ROBIN);
+
+    assert EndpointGroupRegistry.get("search_engines") == searchEngineGroup;
+
+.. note::
+
+    You can create an :api:`Endpoint` with non-default weight using ``withWeight()`` method:
+
+    .. code-block:: java
+
+        // The default weight is 1000.
+        Endpoint endpointWithDefaultWeight = Endpoint.of("foo.com", 8080);
+        Endpoint endpointWithCustomWeight = endpointWithDefaultWeight.withWeight(1500);
+        assert endpointWithDefaultWeight.weight() == 1000;
+        assert endpointWithCustomWeight.weight() == 1500;
+
+
+Connecting to an ``EndpointGroup``
+----------------------------------
+
+Once an :api:`EndpointGroup` is registered, you can use its name in the authority part of a URI:
+
+.. code-block:: java
+
+    // Create an HTTP client that sends requests to the 'search_engines' group.
+    HttpClient client = HttpClient.of("https://group:search_engines/");
+
+    // Send a GET request to each search engine.
+    List<CompletableFuture<? >> futures = new ArrayList<>();
```

Comment:
`<? >>` is intentional(the space after `?`)?

Reply:
No. Was a mistake :sweat_smile: